docs/DeveloperGuide.md

```diff
@@ -1,25 +1,258 @@
+`toString(boolean isQuestion, int offset)` method of a `FlashCard`
+object.
+
+![DG-Implementation-Features-PP-Sequence](./images-dg/PP-Sequence.png?raw=true)
+
+The `offset` parameter specifies the number of characters already
+printed on the line before the flashcard field will be printed.
+The `offset` is used by the `formatResponse()` method to determine
+`usableLength`.
+
+`formatResponse()` places as many words as possible on each line until
+the next word does not fit within the `usableLength` of the current
+line. This word is therefore placed on the next line and the process
+repeats until all the words have been formatted into the response. If
+the  length of a single word exceeds the `usableLength`, the word is
+split across multiple lines to prevent the program from looping
+infinitely as it would never be able to fit the word on any line.
+
+Take note that infinite loops can still occur if
+* formatResponse() is called with offset >= `lineLength` or
+* toString(boolean, int) is called with offset >= `lineLength` - length
+of `label`
+
+#### Design Consideration:
+
+In order to maximize `usableLength`,
+`toString(boolean isQuestion, int offset)` is designed to take
+parameter `offset` instead of hardcoding `offset` to be
+`"2147483647. ".length()` which is the maximum possible index when
+listing flashcards. As a result, flashcards with different number of
+digits in the index will be misaligned when listing flashcards.
+
+### Tags for grouping and searching decks
+The purpose of this feature is to provide a mean to group the decks based on their subjects
+and search for relevant decks related to one or more relevant subjects in a robust way. Each created deck will
+tagged to their respective field.
+
+![DG-Implementation-Features-TagArchitecture](./images-dg/TagFeature-UML.png?raw=true)
+
+The user can also modify to tags of the decks by using tag or untag command, and uses search by tag to find
+a group of decks he/she is interested in.
+
+![DG-Implementation-Features-TagSequence](./images-dg/Tag.png?raw=true)
+
+### Saving to text file (Wayne)
+
+### Game Mode
+
+eCardnomics' quintessential mode. Game Mode can be started from either Normal Mode or Deck Mode. The `start` command
+ is parsed by `NormalParser` (see [Commands](#commands)).
+ 
+Game Mode contains two main components: a storage component, `GameStorage`, and a logic component, `GameEngine`. The
+ former handles all data structures used by Game Mode, and stores the original deck (`originalDeck`), question pool
+  (`deque`), and retest question pool (`retestStore`). The latter executes the main game loop (`runGameLoop()`), and
+   interacts with `GameStorage` on package-private basis; i.e., `GameEngine` and `GameStorage` have full mutual
+    access as if they were a single class. This is one of the main intentional design decisions.
+  
+![DG-Implementation-Features-Game-Mode-Architecture](./images-dg/Game-Mode-Design.png?raw=true "Game Mode
```

Comment:
Maybe you would like to split up this digram into two parts, so that it is less complicated?

Reply:
Changed, thanks!